SHIP-0028: BuildRun cleanup

SaschaSchwarze0 opened the pull request:
Looking for some feedback already especially on the TBDs inside the ship.

SaschaSchwarze0 commented:
/assign ENCALADA
/assign HeavyWombat

SaschaSchwarze0 commented:
/assign qu1queee

gabemontero commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,150 @@
+
+The two TTL fields behaves the same as the [`ttlSecondsAfterFinished` field for Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/). I decided against naming consistency with that feature, in favor of consistency inside our resource. Builds and BuildRuns already have a `timeout` field where we also use a duration while Kubernetes Jobs use an integer to store just se ...
+
+The `successfulLimit` and `failedLimit` fields hold a positive integer number. They define how many successful and failed BuildRuns for a Build can exist.
```
My assumption is that if *either* the TTL or limit count is reached, pruning occurs.

But let's not assume it is obvious (which ever way you are thinking) and state that here, even if you make it clearer in the backend/details section.

By comparison, I would expect that distinction to be clear in the godoc of the API go files vs. having to pick up on that fact in the associated markdown files.

gabemontero reviewed (commented):
a couple of quick thoughts on what you have @SaschaSchwarze0

SaschaSchwarze0 commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,150 @@
+
+Our BuildRun reconciliation logic will need to be extended. [After we fetched the BuildRun](https://github.com/shipwright-io/build/blob/v0.8.0/pkg/reconciler/buildrun/buildrun.go#L82), we will need to add logic that checks if the BuildRun is already completed and check if we need to delete it or requeue for deletion.
+
+TBD When a Build gets modified and any value inside the `retention` section gets initially set or lowered, then ...
```
That's also matching my gut feeling. Added it to the agenda for next Monday to discuss it. https://github.com/shipwright-io/community/issues/62#issuecomment-1041378397

gabemontero replied:
sounds good

qu1queee commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,152 @@
+spec:
+  retention:
+    ttlAfterFailed: duration
+    ttlAfterSuccessful: duration
```
on the wording, `successful` vs `succeeded` . Succeeded inlines with how we do in the `.status.conditions`, see https://github.com/shipwright-io/build/blob/main/docs/buildrun.md#understanding-the-state-of-a-buildrun

SaschaSchwarze0 replied:
Done.

qu1queee commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,152 @@
+
+We can consider to run above logic in a go routine rather than as part of the BuildRun reconciliation. Failures would be ignored which is okay.
+
+When the `ttlAfterFailed` field is set for a BuildRun that failed, or when the `ttlAfterSuccessful` field is set for a BuildRun that succeeded, then we will return a reconciliation result where `Requeue` is set to `true` and `RequeueAfter` is set to `.status.completionTime + TTL - now`.
```
any reasons why you didnt choose the `lasttransitiontime` of our `Succeeded` condition. I believe is more accurate as it based on our controller updates(last state transition)

SaschaSchwarze0 replied:
Makes sense, had forgotten that we take `completionTime` from the TaskRun.

SaschaSchwarze0 replied:
Done.

qu1queee commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,152 @@
+
+### Risks and Mitigations
+
+We will have more logic in our reconciliations. Especially the required `list` calls can become a problem if `successfulLimit`, or `failedLimit` are set to high values and we consequently retrieve a large amount of BuildRuns from the API server. A mitigation would be that we define maximum values for those values (maybe 100).
```
+1, this contradicts the one-shot previous efforts, due to extra reconcile loops.

qu1queee commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,152 @@
+
+None.
+
+## Alternatives
```
Can we consider having a separate controller? I need to spend more time thinking on the pros/cons of this.

SaschaSchwarze0 replied:
Yep, I had that idea as well. Need to think about this again now that we discussed how we handle Build updates.

qu1queee replied:
ok, let me try to brainstorm today/tomorrow, before putting my thoughts here. Thanks.

qu1queee replied:
Adding here my thoughts(already expressed them offline to Sascha):

My concern on not using another controller, is in terms of the reconcile loop increase. My rational in terms of reconciles is the following(based on the worst scenario on ttl):

Assuming 600 completed Buildruns with TTL. 
- Today a single BR reconciles 4-5 times only. => 3000 reconciles
- We increase per BR +1 reconcile(due to the requeue)  => +600 reconciles
- On rollouts of the controller,  assuming all 600 are completed and waiting for a requeue(worst scenario), thats a +600 (we didnt do this before) => +600 reconciles
- In total +1200, a 40% plus. Significant increase.

An alternative for the rollouts would be introducing rate limiting in the controller, see [ratelimiter](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/ratelimiter#RateLimiter) that might help with one of the extra 600.

gabemontero replied:
Some quick "prior art" for reference.  OpenShift build pruning has handled things in "the same" controller / reconciler loop for many years with little to no problems, but OpenShift builds does not have TTL based pruning, only count/total based.

It will be interesting to see as we work the implementation which of the following approaches turn out most palatable:
- count based pruning in the existing controller/reconciler loop, but ttl based in a new controller/reconciler
- both count and ttl based in a new controller/reconciler
- both count and ttl based in the existing controller/reconciler

Seeing this also resulted in the following notion popping into my head:  do we need both TTL and count based pruning out of the gate?  i.e. is there some CodeEngine specific req wanting TTL?  Or is it sufficient to go out with say just count based pruning initially, then wait for feedback on whether that is sufficient?

FWIW, there have been no requests for TTL based pruning that I can recall from OpenShift Build customers (where OCP Builds was first released in 2015).

SaschaSchwarze0 replied:
Thanks @gabemontero for the feedback. The main motivation to have the TTL is related to the standalone BuildRuns where there is no Build (see the other ship from Baran). If there is no Build, there is no count-based pruning.

gabemontero replied:
ah thanks for the clarification @SaschaSchwarze0 ... yeah, I have not had a chance to look at that EP yet, and the title of `One-Off Builds` did not immediately connect with me that it meant "BuildRuns without Builds", but yes it makes sense, that without that organizational aspect of "being under a Build", the count based pruning does not make sense.  Or more precisely, the only option is "global" limits and that is not what is desired.

I'm wrapping up my review of the totality of the latest changes now.  Thanks.

SaschaSchwarze0 commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,152 @@
+
+## Motivation
+
+In Shipwright, we separated the definition of a build and the run of a build into two artifacts, the Build and BuildRun. The BuildRun by its nature is a run-to-completion workload. Once it is completed, its only purpose is to view its status. Retaining the BuildRun for some time makes sense. In case of a successful BuildRun on finds there the digest of the image. For a failed BuildRun, one finds  ...
```
Done.

SaschaSchwarze0 commented:
@qu1queee @gabemontero I added a commit with updates. Those mainly cover two things:

1. Describing the behavior when Build updates happen after BuildRuns were completed.
2. I changed the implementation considerations to use new controllers for the cleanup. Knowing the complexity of the BuildRun reconciler, I feel much better with this approach as it would have meant many additional `if`-style code in the existing one and the increased complexity would have further increased the risk that we oversee certain scenarios.

Please have a look. Thank you.

qu1queee commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -84,26 +84,39 @@ If both TTL and limit are defined, then the BuildRun will get deleted once the f
-- If the number of succeeded BuildRuns is larger than `succeededLimit`, then the oldest succeeded BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
-- If the number of failed BuildRuns is larger than `failedLimit`, then the oldest failed BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+- Create: if the BuildRun is completed (condition `type=Succeeded` set to `"true"` or `"false"`), and if a applicable TTL is set. This is handled for managers becoming the leader.
+- Update: if the BuildRun completed (condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if an applicable TTL is set.
```
to be really precise 

> Condition `type=Succeeded` changes its `.Status` field, from ...

qu1queee commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -133,19 +146,28 @@ Should be available before release v1.0.0
-We will have more logic in our reconciliations. Especially the required `list` calls can become a problem if `succeededLimit`, or `failedLimit` are set to high values and we consequently retrieve a large amount of BuildRuns from the API server. A mitigation would be that we define maximum values for those values (maybe 100).
+We will have more controllers and reconcilers. Consumers may have to increase memory limits in the deployment.
+
+For the support of limits, especially the required `list` calls can become a problem if `succeededLimit`, or `failedLimit` are set to high values and we consequently retrieve a large amount of BuildRuns from the API server and have them in memory. A mitigation would be that we define maximum values for those values (maybe 100).
```
yes, we need this.

qu1queee reviewed (approved):
Approving. Looks neat, moving forward with two separate controllers is ideal, if we would have opt-int for extending the current controllers only, then we might have just started with the limits, and not ttl.

/approve

gabemontero reviewed (approved):
I've got nothing to add @SaschaSchwarze0 .  I see @qu1queee added a couple of comments and while I don't *think* he is asking for changes (instead just making anecdotal comments) I'm not 100% sure, so I'll refrain on adding a lgtm label, along with I see you have @adambkaplan listed as an approver, so I'm taking that as you wanting him to take a pass at this as well before it merges.

SaschaSchwarze0 commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,174 @@
+
+It will watch BuildRuns with those predicate functions:
+
+- Update: if the BuildRun completed (condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if it is related to a Build (standalone BuildRuns when introduced with SHOP 0029 will be ignored).
```
:-)

SaschaSchwarze0 commented:
> I've got nothing to add @SaschaSchwarze0 . I see @qu1queee added a couple of comments and while I don't _think_ he is asking for changes (instead just making anecdotal comments) I'm not 100% sure, so I'll refrain on adding a lgtm label, along with I see you have @adambkaplan listed as an approver, so I'm taking that as you wanting him to take a pass at this as well before it merges.

I added the remaining small changes (status of conditions as mentioned by Enrique, and the SHOP vs SHIP typo found by Raghav). I also change the approvers section. I only initialized them that way. Commits also squashed. Looking for a lgtm. @gabemontero ;-)

adambkaplan commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,152 @@
+
+A successful BuildRun is a BuildRun where the `type=Succeeded` condition is set to `"true"`. A failed BuildRun has the `type=Succeeded` condition is set to `"false"`, and contains therefore for example canceled BuildRuns as well.
+
+The two TTL fields behaves the same as the [`ttlSecondsAfterFinished` field for Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/). I decided against naming consistency with that feature, in favor of consistency inside our resource. Builds and BuildRuns already have a `timeout` field where we also use a duration while Kubernetes Jobs use an integer to store just se ...
```
+1 to using `Duration`.

I also like having separate TTLs for success and failure - a failed BuildRun may need a longer TTL so the dev team can inspect logs.

adambkaplan commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,175 @@
+- Create: if the BuildRun is completed (`status` of condition `type=Succeeded` set to `"true"` or `"false"`), and if a applicable TTL is set. This is handled for managers becoming the leader.
+- Update: if the BuildRun completed (`status` of condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if an applicable TTL is set.
+
+In the `Reconcile` function, we will check the BuildRun: When the `ttlAfterFailed` field is set for a BuildRun that failed, or when the `ttlAfterSucceeded` field is set for a BuildRun that succeeded, then we will calculate `.status.conditions[type=Succeeded].lastTransitionTime + TTL - now`. If the value is smaller than or equal zero, then we will delete the BuildRun. Otherwise, we will return a r ...
```
+1

adambkaplan commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,175 @@
+
+It will watch BuildRuns with those predicate functions:
+
+- Update: if the BuildRun completed (`status` of condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if it is related to a Build (standalone BuildRuns when introduced with SHIP 0029 will be ignored).
```
+1 to this as well - standalone `BuildRun` objects can be considered "orphans" and should only use TTL for cleanup.

adambkaplan commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,175 @@
+
+We will have more controllers and reconcilers. Consumers may have to increase memory limits in the deployment.
+
+For the support of limits, especially the required `list` calls can become a problem if `succeededLimit`, or `failedLimit` are set to high values and we have to retrieve a large amount of BuildRuns from the API server and have them in memory. A mitigation would be that we define maximum values for those values (maybe 100).
```
Perhaps we enforce an upper bound on succeeded + failed limits in the main `Build` controller and/or an additional webhook? Something to consider as a future enhancement.

adambkaplan commented on ships/0028-buildrun-cleanup.md:
```diff
@@ -0,0 +1,175 @@
+
+## Drawbacks
+
+None.
```
More of a nit - the drawbacks are addressed in the "Risks and Mitigations" section.

adambkaplan reviewed (approved):
/lgtm

This is a great proposal @SaschaSchwarze0! I've noted a few follow up areas in the proposal that we may want to consider. Otherwise this looks good.
